src/network/core/http_curl.cpp

```diff
@@ -12,216 +12,165 @@
-	curl_easy_setopt(this->curl, CURLOPT_USERAGENT, user_agent.c_str());
-	curl_easy_setopt(this->curl, CURLOPT_FOLLOWLOCATION, 1L);
-	curl_easy_setopt(this->curl, CURLOPT_MAXREDIRS, 5L);
-
-	/* Give the connection about 10 seconds to complete. */
-	curl_easy_setopt(this->curl, CURLOPT_CONNECTTIMEOUT, 10L);
-
-	/* Set a buffer of 100KiB, as the default of 16KiB seems a bit small. */
-	curl_easy_setopt(this->curl, CURLOPT_BUFFERSIZE, 100L * 1024L);
-
-	/* Fail our call if we don't receive a 2XX return value. */
-	curl_easy_setopt(this->curl, CURLOPT_FAILONERROR, 1L);
-
-	/* Create a multi-handle so we can do the call async. */
-	this->multi_handle = curl_multi_init();
-	curl_multi_add_handle(this->multi_handle, this->curl);
-
-	/* Trigger it for the first time so it becomes active. */
-	int still_running;
-	curl_multi_perform(this->multi_handle, &still_running);
 }
 
-/**
- * Poll and process the HTTP request/response.
- *
- * @return True iff the request is done; no call to Receive() should be done after it returns true.
- */
-bool NetworkHTTPRequest::Receive()
+void HttpThread()
 {
-	int still_running = 0;
-
-	/* Check for as long as there is activity on the socket, but once in a while return.
-	 * This allows the GUI to always update, even on really fast downloads. */
-	for (int count = 0; count < 100; count++) {
-		/* Check if there was activity in the multi-handle. */
-		int numfds;
-		curl_multi_wait(this->multi_handle, NULL, 0, 0, &numfds);
-		if (numfds == 0) return false;
-
-		/* Let CURL process the activity. */
-		curl_multi_perform(this->multi_handle, &still_running);
-		if (still_running == 0) break;
-	}
+	CURL *curl = curl_easy_init();
+	assert(curl != nullptr);
 
-	/* The download is still pending (so the count is reached). Update GUI. */
-	if (still_running != 0) return false;
-
-	/* The request is done; check the result and close up. */
-	int msgq;
-	CURLMsg *msg = curl_multi_info_read(this->multi_handle, &msgq);
-	/* We can safely assume this returns something, as otherwise the multi-handle wouldn't be empty. */
-	assert(msg != nullptr);
-	assert(msg->msg == CURLMSG_DONE);
-
-	CURLcode res = msg->data.result;
-	if (res == CURLE_OK) {
-		Debug(net, 1, "HTTP request succeeded");
-		this->callback->OnReceiveData(nullptr, 0);
-	} else {
-		Debug(net, 0, "HTTP request failed: {}", curl_easy_strerror(res));
-		this->callback->OnFailure();
-	}
+	for (;;) {
+		std::unique_lock<std::mutex> lock(_http_mutex);
 
-	return true;
-}
+		/* Wait for a new request. */
+		while (_http_requests.empty() && !_http_thread_exit) {
+			_http_cv.wait(lock);
+		}
+		if (_http_thread_exit) break;
 
-/**
- * Destructor of the HTTP request.
- *
- * Makes sure all handlers are closed, and all memory is free'd.
- */
-NetworkHTTPRequest::~NetworkHTTPRequest()
-{
-	if (this->curl) {
-		curl_multi_remove_handle(this->multi_handle, this->curl);
-		curl_multi_cleanup(this->multi_handle);
+		std::shared_ptr<NetworkHTTPRequest> request = _http_requests.front();
```

Comment:
I think we do not need `shared_ptr`, but can use `unique_ptr` if `move`-ing from `front()`:
```suggestion
		std::unique_ptr<NetworkHTTPRequest> request = std::move(_http_requests.front());
```
The other `shared`s in this PR also need to become `unique` for this to work.